sips/sip-021/sip-021-nakamoto.md

```diff
@@ -0,0 +1,1040 @@
+* The stacking minimum is abolished.  If there is at least one stacker that is
+  able to claim one reward slot, then PoX activates.
+* The `.pox-4` contract will expose each reward cycle's full reward set,
+  including the signing keys for each stacker, via a public function.
+Internally, the Stacks node will call a _private_ function to load the reward
+set into the `.pox-4` data space after it identifies the PoX anchor block.  This
+is required for other dependent SIPs, such as the sBTC SIP.
+
+## Dynamic Runtime Budgets
+
+The time between cryptographic sortitions (and thus tenure changes) depends on
+the time between two consecutive burnchain blocks.  This can be highly variable,
+which complicates the task of sustaining a predictable transaction confirmation
+latency while also preventing a malicious miner from spamming the network with
+too many high-resource transactions.
+
+Today, each miner receives a tenure block budget, which places hard limits on
+how much CPU, RAM, and I/O their block can consume when evaluated.  In this
+proposal, each miner begins its tenure with a fixed budget, but Stackers may opt
+to increase that budget through a vote.  This is done to enable the miner to
+continue to produce blocks if the next burnchain block is late.
+
+To achieve this, Stackers each keep track of the elapsed wall-clock time since
+the start of the tenure.  Once the expected tenure time has passed (e.g. 10
+minutes), they vote to grant the miner an additional tenure execution budget.
+This is achieved by having Stackers collectively produce a `TenureExtension`
+Stacks transaction, which the miner would include into one of their blocks.
+Once this transaction has been processed, the tenure block budget expands.
+
+## Support for WSTS Schnorr Signatures
+
+This proposal calls for adding support for WSTS Schnorr signatures as a
+transaction authorization mode to Stacks.  This enables Stackers to collectively
+sign a Stacks transaction.  The specification for the transaction authorization
+is as follows:
+
+* The spending condition's hash mode must be **0x04**.
+* An additional spending condition variant specific to single-key Schnorr
+  signatures, which can only be used if the hash mode is 0x04.  The spending
+condition variant is encoded as follows:
+  * A 65-byte **Schnorr signature**, consisting of:
+    * The 33-byte signed x-coordinate of the signature curve point
+    * The 32-byte scalar derived from the hash of the generator raised to the
+      _k_th power and the message
+  * A 33-byte **compressed secp256k1 public key**, consisting of its signed
+    x-coordinate
+
+To verify the transaction signature, the node both verifies that the public
+key's hash160 is equal to the spending condition's 20-byte hash, and verifies
+that the Schnorr signature over the `presign-sighash` value of the transaction
+was generated by corresponding private key.
+
+The address of a principal who signs a transaction with the WSTS Schnorr
```

Comment:
I'm on the fence about this, tbh.  On the one hand, keeping the signing algorithm a secret is a privacy feature, much like how a p2sh or p2tr address in Bitcoin does not reveal anything about the spending conditions.  This goes one step further: the user can choose whether or not to sign with either ECDSA or Schnorr, but the address would be the same, thereby granting the user a degree of cryptographic agility.  This is because with ECDSA, we recover the original public key from the signature, and like in the described the Schnorr scheme, we verify that its hash160 equals the spending condition's 20-byte hash.

On the other hand, this means wallets would need to choose the signing algorithm when signing a transaction.  They'd want to use [RFC 6979](https://datatracker.ietf.org/doc/html/rfc6979) for their _k_ calculation, but I'm pretty sure they all do this these days.

Reply:
This might be a @xoloki question.  Is there a risk of supporting two separate signing methods (ECDSA and Schnorr) for the same address, such that the user can choose whichever method they want at any time?  Namely, I'm concerned about the user becoming a signing oracle.